File: output/GPT3.5-Turbo/xCodeEval_4000/Java/Go/dff2744b9701a10438fa6c925781d1b5.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
)

type Graph struct {
	vertexCount int
	edgeCount   int
	firstOut    []int
	from        []int
	to          []int
	nextOut     []int
	flags       []int
}

type Edge struct {
	id int
}

type IntList []int

func (list IntList) Len() int {
	return len(list)
}

func (list IntList) Less(i, j int) bool {
	return list[i] < list[j]
}

func (list IntList) Swap(i, j int) {
	list[i], list[j] = list[j], list[i]
}

func (list *IntList) Push(x interface{}) {
	*list = append(*list, x.(int))
}

func (list *IntList) Pop() interface{} {
	old := *list
	n := len(old)
	x := old[n-1]
	*list = old[0 : n-1]
	return x
}

func (graph *Graph) AddEdge(fromID, toID int) {
	graph.ensureEdgeCapacity(graph.edgeCount + 1)
	if graph.firstOut[fromID] != -1 {
		graph.nextOut[graph.edgeCount] = graph.firstOut[fromID]
	} else {
		graph.nextOut[graph.edgeCount] = -1
	}
	graph.firstOut[fromID] = graph.edgeCount
	graph.from[graph.edgeCount] = fromID
	graph.to[graph.edgeCount] = toID
	graph.flags[graph.edgeCount] = 0
	graph.edgeCount++
}

func (graph *Graph) ensureEdgeCapacity(size int) {
	if len(graph.from) < size {
		newSize := max(size, 2*len(graph.from))
		newFrom := make([]int, newSize)
		newTo := make([]int, newSize)
		newNextOut := make([]int, newSize)
		newFlags := make([]int, newSize)
		copy(newFrom, graph.from)
		copy(newTo, graph.to)
		copy(newNextOut, graph.nextOut)
		copy(newFlags, graph.flags)
		graph.from = newFrom
		graph.to = newTo
		graph.nextOut = newNextOut
		graph.flags = newFlags
	}
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func solve(testNumber int, in *InputReader, out io.Writer) {
	n := in.readInt()
	k := in.readInt()
	a := make([]int, n)
	for i := 0; i < n; i++ {
		a[i] = in.readInt()
	}
	u := make([]int, n-1)
	v := make([]int, n-1)
	for i := 0; i < n-1; i++ {
		u[i] = in.readInt()
		v[i] = in.readInt()
	}
	graph := createGraph(n, u, v)
	order := make(IntList, n)
	for i := 0; i < n; i++ {
		order[i] = i
	}
	sort.Sort(sort.Reverse(order))
	left := k
	right := n
	weight := make([]int, n)
	degree := make([]int, n)
	for i := 0; i < n-1; i++ {
		degree[u[i]]++
		degree[v[i]]++
	}
	remaining := make([]int, n)
	queue := make([]int, n)
	path := make([]int, n)
	for left < right {
		middle := (left + right) >> 1
		threshold := a[order[middle-1]]
		for i := 0; i < n; i++ {
			if a[i] >= threshold {
				weight[i] = 1
			} else {
				weight[i] = 0
			}
		}
		copy(remaining, degree)
		size := 0
		for i := 0; i < n; i++ {
			if degree[i] == 1 && weight[i] > 0 {
				queue[size] = i
				size++
			}
		}
		for i := 0; i < size; i++ {
			current := queue[i]
			for j := graph.firstOut[current]; j != -1; j = graph.nextOut[j] {
				next := graph.to[j]
				if weight[next] != 0 {
					weight[next] += weight[current]
					weight[current] = 0
					remaining[next]--
					if remaining[next] == 1 {
						queue[size] = next
						size++
					}
				}
			}
		}
		result := dfs(0, -1, graph, weight, path)
		if result >= k {
			right = middle
		} else {
			left = middle + 1
		}
	}
	fmt.Fprintln(out, a[order[left-1]])
}

func dfs(current, last int, graph *Graph, weight, path []int) int {
	result := 0
	max := 0
	second := 0
	for i := graph.firstOut[current]; i != -1; i = graph.nextOut[i] {
		next := graph.to[i]
		if next == last {
			continue
		}
		result = max(result, dfs(next, current, graph, weight, path))
		if path[next] > max {
			second = max
			max = path[next]
		} else {
			second = max(second, path[next])
		}
	}
	if weight[current] != 0 {
		path[current] = weight[current] + max
		result = max(result, max+second+weight[current])
	} else {
		path[current] = 0
	}
	return result
}

func createGraph(vertexCount int, from, to []int) *Graph {
	graph := &Graph{vertexCount: vertexCount, firstOut: make([]int, vertexCount)}
	for i := 0; i < vertexCount; i++ {
		graph.firstOut[i] = -1
	}
	graph.from = make([]int, 2*len(from))
	graph.to = make([]int, 2*len(from))
	graph.nextOut = make([]int, 2*len(from))
	graph.flags = make([]int, 2*len(from))
	for i := 0; i < len(from); i++ {
		graph.AddEdge(from[i], to[i])
		graph.AddEdge(to[i], from[i])
	}
	return graph
}

type InputReader struct {
	reader *bufio.Reader
}

func NewInputReader(reader io.Reader) *InputReader {
	return &InputReader{reader: bufio.NewReader(reader)}
}

func (in *InputReader) readInt() int {
	var sign, res int
	var err error
	for {
		sign = 1
		res = 0
		var flag bool
		var c byte
		for {
			c, err = in.reader.ReadByte()
			if err != nil {
				return res * sign
			}
			if c >= '0' && c <= '9' {
				res = int(c - '0')
				break
			}
			if c == '-' {
				sign = -1
				break
			}
			flag = true
		}
		if flag {
			break
		}
		for {
			c, err = in.reader.ReadByte()
			if err != nil {
				return res * sign
			}
			if c < '0' || c > '9' {
				return res * sign
			}
			res = res*10 + int(c-'0')
		}
		break
	}
	for {
		c, err := in.reader.ReadByte()
		if err != nil || c == '\n' || c == ' ' {
			return res * sign
		}
	}
}

func main() {
	in := NewInputReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()
	solve(1, in, out)
}

File: output/GPT4o-Mini/xCodeEval_remove_4000/Java/C/dff2744b9701a10438fa6c925781d1b5.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

typedef struct {
    int *data;
    int size;
} IntArray;

typedef struct {
    int *firstOutbound;
    int *nextOutbound;
    int *from;
    int *to;
    long *weight;
    long *capacity;
    int vertexCount;
    int edgeCount;
} Graph;

typedef struct {
    int *degree;
    int *weight;
    int *path;
    Graph graph;
} TaskD;

void initIntArray(IntArray *arr, int size) {
    arr->data = (int *)malloc(size * sizeof(int));
    arr->size = size;
}

void freeIntArray(IntArray *arr) {
    free(arr->data);
}

void addEdge(Graph *graph, int fromID, int toID, long weight, long capacity) {
    if (graph->edgeCount >= graph->vertexCount) {
        graph->vertexCount *= 2;
        graph->from = realloc(graph->from, graph->vertexCount * sizeof(int));
        graph->to = realloc(graph->to, graph->vertexCount * sizeof(int));
        graph->weight = realloc(graph->weight, graph->vertexCount * sizeof(long));
        graph->capacity = realloc(graph->capacity, graph->vertexCount * sizeof(long));
    }
    graph->from[graph->edgeCount] = fromID;
    graph->to[graph->edgeCount] = toID;
    graph->weight[graph->edgeCount] = weight;
    graph->capacity[graph->edgeCount] = capacity;
    graph->edgeCount++;
}

void dfs(TaskD *task, int current, int last, int *result) {
    int max = 0, second = 0;
    for (int i = 0; i < task->graph.edgeCount; i++) {
        if (task->graph.from[i] == current && task->graph.to[i] != last) {
            int next = task->graph.to[i];
            dfs(task, next, current, result);
            if (task->path[next] > max) {
                second = max;
                max = task->path[next];
            } else {
                second = (task->path[next] > second) ? task->path[next] : second;
            }
        }
    }
    if (task->weight[current] != 0) {
        task->path[current] = task->weight[current] + max;
        *result = (*result > max + second + task->weight[current]) ? *result : (max + second + task->weight[current]);
    } else {
        task->path[current] = 0;
    }
}

void solve(TaskD *task, int n, int k, int *a, int *u, int *v) {
    task->graph.vertexCount = n;
    task->graph.edgeCount = 0;
    task->graph.from = (int *)malloc((n - 1) * sizeof(int));
    task->graph.to = (int *)malloc((n - 1) * sizeof(int));
    task->graph.weight = (long *)malloc((n - 1) * sizeof(long));
    task->graph.capacity = (long *)malloc((n - 1) * sizeof(long));
    task->degree = (int *)calloc(n, sizeof(int));
    task->weight = (int *)malloc(n * sizeof(int));
    task->path = (int *)malloc(n * sizeof(int));

    for (int i = 0; i < n - 1; i++) {
        task->degree[u[i]]++;
        task->degree[v[i]]++;
        addEdge(&task->graph, u[i], v[i], 1, 1);
    }

    int left = k, right = n;
    while (left < right) {
        int middle = (left + right) / 2;
        int threshold = a[middle - 1];
        memset(task->weight, 0, n * sizeof(int));
        for (int i = 0; i < n; i++) {
            task->weight[i] = (a[i] >= threshold) ? 1 : 0;
        }
        int result = 0;
        dfs(task, 0, -1, &result);
        if (result >= k) {
            right = middle;
        } else {
            left = middle + 1;
        }
    }
    printf("%d\n", a[left - 1]);

    free(task->graph.from);
    free(task->graph.to);
    free(task->graph.weight);
    free(task->graph.capacity);
    free(task->degree);
    free(task->weight);
    free(task->path);
}

int main() {
    int n, k;
    scanf("%d %d", &n, &k);
    int *a = (int *)malloc(n * sizeof(int));
    int *u = (int *)malloc((n - 1) * sizeof(int));
    int *v = (int *)malloc((n - 1) * sizeof(int));

    for (int i = 0; i < n; i++) {
        scanf("%d", &a[i]);
    }
    for (int i = 0; i < n - 1; i++) {
        scanf("%d %d", &u[i], &v[i]);
    }

    TaskD task;
    solve(&task, n, k, a, u, v);

    free(a);
    free(u);
    free(v);
    return 0;
}
